src/value.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2014 The Prometheus Authors
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::RwLock;
+
+use protobuf::RepeatedField;
+
+use proto::{LabelPair, Metric, Counter, Gauge, Untyped};
+use desc::Desc;
+use errors::{Result, Error};
+
+pub enum ValueType {
+    Counter,
+    Gauge,
+    Untyped,
+}
+
+/// `Value` is a generic metric for Counter, Gauge and Untyped.
+pub struct Value {
+    pub desc: Desc,
+    // TODO: like prometheus client go, use atomic u64.
+    pub val: RwLock<f64>,
+    pub val_type: ValueType,
+    pub label_pairs: Vec<LabelPair>,
+}
+
+impl Value {
+    pub fn new(desc: Desc,
+               value_type: ValueType,
+               val: f64,
+               label_values: Vec<String>)
+               -> Result<Value> {
+        if desc.variable_labels.len() != label_values.len() {
+            return Err(Error::InconsistentCardinality);
+        }
+
+        let label_pairs = make_label_pairs(&desc, label_values);
+
+        Ok(Value {
+            desc: desc,
+            val: RwLock::new(val),
+            val_type: value_type,
+            label_pairs: label_pairs,
+        })
+    }
+
+    #[inline]
+    pub fn set(&self, val: f64) {
+        *self.val.write().unwrap() = val;
+    }
+
+    #[inline]
+    pub fn get(&self) -> f64 {
+        *self.val.read().unwrap()
+    }
+
+    #[inline]
+    pub fn inc(&self) {
+        self.add(1.0);
+    }
+
+    #[inline]
+    pub fn dec(&self) {
+        self.add(-1.0);
+    }
+
+    #[inline]
+    pub fn add(&self, val: f64) {
```

Comment:
I think this can slightly improved in some ways, either:
 1. making this closer to the [Add trait](https://doc.rust-lang.org/std/ops/trait.Add.html) by making it return something (perhaps a `Result<f64>`?)
 2. turning it into a `add_assign` which mutate in place, as per [AddAsign](https://doc.rust-lang.org/std/ops/trait.AddAssign.html)